### PR TITLE
refactor: remove browserPaths in favor of Registry class

### LINKED ISSUE #5278: chore: follow-up with ffmpeg downloads to registry
- [x] include license to FFMPEG archive
- [ ] code cleanup: move `packagePath` var to `playwrightOptions`

### PR DESCRIPTION
**Pull request by aslushnikov**
This patch introduces a new Registry class that incapsulates
all logic regarding browsers and their paths.

Fixes #5278

### CODE REVIEW

File: src/utils/registry.ts [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+import * as fs from 'fs';
+import * as os from 'os';
+import * as path from 'path';
+import * as util from 'util';
+import { getUbuntuVersionSync } from './ubuntuVersion';
+import { assert, getFromENV } from './utils';
+
+export type BrowserName = 'chromium'|'webkit'|'firefox'|'ffmpeg';
+export type BrowserPlatform = 'win32'|'win64'|'mac10.13'|'mac10.14'|'mac10.15'|'mac11'|'mac11-arm64'|'ubuntu18.04'|'ubuntu20.04';
+export type BrowserDescriptor = {
```
Comment (dgozman) [inline src/utils/registry.ts:28]: No need to export?

File: src/utils/registry.ts [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+import { execSync } from 'child_process';
+import * as fs from 'fs';
+import * as os from 'os';
+import * as path from 'path';
+import * as util from 'util';
+import { getUbuntuVersionSync } from './ubuntuVersion';
+import { assert, getFromENV } from './utils';
+
+export type BrowserName = 'chromium'|'webkit'|'firefox'|'ffmpeg';
+export type BrowserPlatform = 'win32'|'win64'|'mac10.13'|'mac10.14'|'mac10.15'|'mac11'|'mac11-arm64'|'ubuntu18.04'|'ubuntu20.04';
```
Comment (dgozman) [inline src/utils/registry.ts:27]: No need to export?

File: src/utils/registry.ts [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    'mac10.14': '%s/builds/ffmpeg/%s/ffmpeg-mac.zip',
+    'mac10.15': '%s/builds/ffmpeg/%s/ffmpeg-mac.zip',
+    'mac11': '%s/builds/ffmpeg/%s/ffmpeg-mac.zip',
+    'mac11-arm64': '%s/builds/ffmpeg/%s/ffmpeg-mac.zip',
+    'win32': '%s/builds/ffmpeg/%s/ffmpeg-win32.zip',
+    'win64': '%s/builds/ffmpeg/%s/ffmpeg-win64.zip',
+  },
+};
+
+export const allBrowserNames: BrowserName[] = Object.keys(DOWNLOAD_URLS) as BrowserName[];
```
Comment (dgozman) [inline src/utils/registry.ts:128]: I'd rather move this next to BrowserName definition and hardcode the list.

File: src/utils/registry.ts [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+export class Registry {
+  private _descriptors: BrowserDescriptor[];
+
+  constructor(packagePath: string) {
+    const browsersJSON = JSON.parse(fs.readFileSync(path.join(packagePath, 'browsers.json'), 'utf8'));
+    this._descriptors = browsersJSON['browsers'];
+  }
+
+  _browser(browserName: BrowserName): BrowserDescriptor {
```
Comment (dgozman) [inline src/utils/registry.ts:186]: private

File: src/utils/registry.ts [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+      return [path.join(browserDirectory, 'firefox')];
+    if (browserName === 'webkit')
+      return [browserDirectory];
+    return [];
+  }
+
+  executablePath(browserName: BrowserName): string | undefined {
+    const browserDirectory = this.browserDirectory(browserName);
+    const tokens = EXECUTABLE_PATHS[browserName][hostPlatform];
+    return tokens ? path.join(browserDirectory, ...tokens) : undefined;
```
Comment (dgozman) [inline src/utils/registry.ts:235]: Why do we support undefined here? Is there a usecase?

File: src/install/installer.ts [hunk 6]
```diff
@@ -55,34 +55,38 @@ export async function installBrowsersWithProgressBar(packagePath: string, browse
...
-      const browsersToDownload = await readBrowsersToDownload(linkTarget);
-      for (const browser of browsersToDownload) {
-        const usedBrowserPath = browserPaths.browserDirectory(browsersPath, browser);
-        const browserRevision = parseInt(browser.revision, 10);
+      const linkRegistry = new Registry(linkTarget);
+      for (const browserName of allBrowserNames) {
+        if (!linkRegistry.shouldDownload(browserName))
+          continue;
+        const usedBrowserPath = linkRegistry.browserDirectory(browserName);
+        const browserRevision = parseInt(linkRegistry.revision(browserName), 10);
```
Comment (dgozman) [inline src/install/installer.ts:82]: Let's move parseInt call to the registry?

File: src/install/installer.ts [hunk 7]
```diff
@@ -91,37 +95,29 @@ async function validateCache(packagePath: string, browsersPath: string, linksDir
...
     browserFetcher.logPolitely('Removing unused browser at ' + directory);
     await removeFolderAsync(directory).catch(e => {});
   }
 
   // 3. Install missing browsers for this package.
-  const myBrowsersToDownload = await readBrowsersToDownload(packagePath, browserNames);
-  for (const browser of myBrowsersToDownload) {
-    await browserFetcher.downloadBrowserWithProgressBar(browsersPath, browser).catch(e => {
-      throw new Error(`Failed to download ${browser.name}, caused by\n${e.stack}`);
+  const mainRegistry = new Registry(PACKAGE_PATH);
```
Comment (dgozman) [inline src/install/installer.ts:109]: nit: mainRegistry -> myRegistry/thisRegistry?

File: src/utils/registry.ts [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+})();
+
+export const registryDirectory = (() => {
+  const envDefined = getFromENV('PLAYWRIGHT_BROWSERS_PATH');
+  if (envDefined === '0')
+    return path.join(__dirname, '..', '..', '.local-browsers');
+  if (envDefined)
+    return path.resolve(process.cwd(), envDefined);
+  return path.join(cacheDirectory(), 'ms-playwright');
+})();
```
Comment (dgozman) [inline src/utils/registry.ts:167]: Let's always do `path.resolve` in this method, to hopefully fix https://github.com/microsoft/playwright/issues/4867?

**Review by dgozman — APPROVED** [review]

File: src/utils/registry.ts (hunk 1 above)
Comment (aslushnikov) [reply src/utils/registry.ts:28]: Done

File: src/utils/registry.ts (hunk 2 above)
Comment (aslushnikov) [reply src/utils/registry.ts:27]: Done

File: src/utils/registry.ts (hunk 3 above)
Comment (aslushnikov) [reply src/utils/registry.ts:128]: sure

File: src/utils/registry.ts (hunk 4 above)
Comment (aslushnikov) [reply src/utils/registry.ts]: This got inlined since it shouldn't actually throw in `shouldDownload()` call

File: src/utils/registry.ts (hunk 5 above)
Comment (aslushnikov) [reply src/utils/registry.ts:231]: playwright-core might return undefined

File: src/install/installer.ts (hunk 6 above)
Comment (aslushnikov) [reply src/install/installer.ts:82]: done

File: src/install/installer.ts (hunk 7 above)
Comment (aslushnikov) [reply src/install/installer.ts:109]: I like `myRegistry`: I had these var names back in the days when I was just starting programming

File: src/utils/registry.ts (hunk 8 above)
Comment (aslushnikov) [reply src/utils/registry.ts:167]: done

**Pull request merged**